src/transaction.js

```diff
@@ -191,33 +193,71 @@ Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashT
   typeForce('Number', hashType)
 
   assert(inIndex >= 0, 'Invalid vin index')
-  assert(inIndex < this.ins.length, 'Invalid vin index')
+
+  // https://github.com/bitcoin/bitcoin/blob/master/src/test/sighash_tests.cpp#L29
+  if (inIndex >= this.ins.length) return one
 
   var txTmp = this.clone()
+
+  // in case concatenating two scripts ends up with two codeseparators,
+  // or an extra one at the end, this prevents all those possible incompatibilities.
   var hashScript = prevOutScript.without(opcodes.OP_CODESEPARATOR)
+  var i
 
-  // Blank out other inputs' signatures
-  txTmp.ins.forEach(function (txIn) {
-    txIn.script = Script.EMPTY
-  })
+  // blank out other inputs' signatures
+  txTmp.ins.forEach(function (input) { input.script = Script.EMPTY })
   txTmp.ins[inIndex].script = hashScript
 
-  var hashTypeModifier = hashType & 0x1f
+  // blank out some of the inputs
+  if ((hashType & 0x1f) === Transaction.SIGHASH_NONE) {
+    // wildcard payee
+    txTmp.outs = []
+
+    // let the others update at will
+    txTmp.ins.forEach(function (input, i) {
+      if (i !== inIndex) {
+        input.sequence = 0
+      }
+    })
+
+  } else if ((hashType & 0x1f) === Transaction.SIGHASH_SINGLE) {
+    var nOut = inIndex
+
+    // only lock-in the txOut payee at same index as txIn
+    // https://github.com/bitcoin/bitcoin/blob/master/src/test/sighash_tests.cpp#L60
```

Comment:
Will do
On 4 May 2015 3:03 am, "Wei Lu" <notifications@github.com> wrote:

> In src/transaction.js
> <https://github.com/bitcoinjs/bitcoinjs-lib/pull/400#discussion_r29557849>
> :
>
> > +  if ((hashType & 0x1f) === Transaction.SIGHASH_NONE) {
> > +    // wildcard payee
> > +    txTmp.outs = []
> > +
> > +    // let the others update at will
> > +    txTmp.ins.forEach(function (input, i) {
> > +      if (i !== inIndex) {
> > +        input.sequence = 0
> > +      }
> > +    })
> > +
> > +  } else if ((hashType & 0x1f) === Transaction.SIGHASH_SINGLE) {
> > +    var nOut = inIndex
> > +
> > +    // only lock-in the txOut payee at same index as txIn
> > +    // https://github.com/bitcoin/bitcoin/blob/master/src/test/sighash_tests.cpp#L60
>
> Link to the sha instead of master:
> https://github.com/bitcoin/bitcoin/blob/90c37bc16c3416f133fb15b726873960e00f2262/src/test/sighash_tests.cpp#L60
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoinjs/bitcoinjs-lib/pull/400/files#r29557849>.
>